File: Java/src/main/java/at/tugraz/networkadapter/MachineInfo.java
package at.tugraz.networkadapter;

import at.tugraz.util.NodeType;

import java.util.ArrayList;

public class MachineInfo {
    private String ip;
    private int port;
    private MachineInfo root;
    private int nrow;
    private int nrowLeaf;
    private ArrayList<MachineInfo> leaves;

    public MachineInfo(String ip, int port, int nrow) {
        this.leaves = new ArrayList<>();
        this.ip = ip;
        this.port = port;
        this.nrow = nrow;
        this.nrowLeaf = 0;
    }

    public String getIp() {
        return ip;
    }

    public void setIp(String ip) {
        this.ip = ip;
    }

    public int getPort() {
        return port;
    }

    public void setPort(int port) {
        this.port = port;
    }

    public int getNrow() {
        return nrow;
    }

    public void setNrow(int nrow) {
        this.nrow = nrow;
    }

    public int getNrowLeaf() {
        return nrowLeaf;
    }

    public void setNrowLeaf(int nrowLeaf) {
        this.nrowLeaf = nrowLeaf;
    }

    public void addLeaf(MachineInfo leaf){
        this.leaves.add(leaf);
    }

    public void setRoot(MachineInfo root) {
        this.root = root;
    }

    public MachineInfo getRoot() {
        return root;
    }

    public ArrayList<MachineInfo> getLeaves() {
        return leaves;
    }

    public void setLeaves(ArrayList<MachineInfo> leaves) {
        this.leaves = leaves;
    }

    public NodeType getNodeType() {
        if (root == null)
            return NodeType.ROOT;
        else if (leaves.size() == 0)
            return NodeType.LEAF;
        else
            return NodeType.MIDDLE;
    }

    public int getTotalNRow(){
        return nrow + nrowLeaf;
    }
}

File: src/main/cpp/util/source/MachineInfo.cpp
#include "MachineInfo.h"

MachineInfo::MachineInfo() {
    this->root = nullptr;
}

MachineInfo::~MachineInfo() {}

MachineInfo::MachineInfo(const string &ip, int port, int nrow){
    this->ip = ip;
    this->port = port;
    this->nrow = nrow;
    this->nrowLeaf = 0;
}

NodeType MachineInfo::getNodeType() {
   if (root == nullptr)
        return ROOT;
    else if (leaves.empty())
        return LEAF;
    else
        return MIDDLE;
}

int MachineInfo::getTotalNRow() {
    return nrow + nrowLeaf;
}

const string &MachineInfo::getIp() const {
    return ip;
}

void MachineInfo::setIp(const string &ip) {
    MachineInfo::ip = ip;
}

int MachineInfo::getPort() const {
    return port;
}

void MachineInfo::setPort(int port) {
    MachineInfo::port = port;
}

MachineInfo *MachineInfo::getRoot() const {
    return root;
}

void MachineInfo::setRoot(MachineInfo *root) {
    MachineInfo::root = root;
}

int MachineInfo::getNrow() const {
    return nrow;
}

void MachineInfo::setNrow(int nrow) {
    MachineInfo::nrow = nrow;
}

int MachineInfo::getNrowLeaf() const {
    return nrowLeaf;
}

void MachineInfo::setNrowLeaf(int nrowLeaf) {
    MachineInfo::nrowLeaf = nrowLeaf;
}

const vector<MachineInfo *> &MachineInfo::getLeaves() const {
    return leaves;
}

void MachineInfo::setLeaves(const vector<MachineInfo *> &leaves) {
    MachineInfo::leaves = leaves;
}

void MachineInfo::addLeaf(MachineInfo * leaf) {
    this->leaves.push_back(leaf);
}
